cloudify_cli/utils.py

```diff
@@ -404,3 +409,78 @@ def print_dict(keys_dict, logger):
 def get_dict_from_yaml(yaml_path):
     with open(yaml_path) as f:
         return yaml.load(f, yaml.Loader)
+
+
+def wait_for_blueprint_upload(client, blueprint_id, timeout=900,
+                              show_error_traceback=False):
+
+    def _handle_errors():
+        if blueprint['state'] in BlueprintUploadState.FAILED_STATES:
+            error_msg = '{error_type} blueprint: {description}.'.format(
+                error_type=blueprint['state'].capitalize().replace('_', ' '),
+                description=blueprint['error']
+            )
+            if show_error_traceback:
+                error_msg += '\nError traceback: {}'.format(
+                    blueprint['error_traceback'])
+            raise CloudifyCliError(error_msg)
+
+    blueprint = client.blueprints.get(blueprint_id)
+
+    # if blueprint upload already ended - return without waiting
+    if blueprint['state'] in BlueprintUploadState.END_STATES:
+        _handle_errors()
+        return blueprint
+
+    if timeout is not None:
+        deadline = time.time() + timeout
+
+    execution = [
+        ex for ex in client.executions.list(workflow_id='upload_blueprint')
+        if ex['parameters']['blueprint_id'] == blueprint_id
+    ][0]
+    events_fetcher = ExecutionEventsFetcher(client,
+                                            execution.id,
+                                            include_logs=True)
+
+    # Poll for execution status and execution logs, until execution ends
+    # and we receive an event of type in WORKFLOW_END_TYPES
+    upload_ended = False
+    events_watcher = EventsWatcher(get_events_logger(None))
+
+    # did we already see the execution status change, and are only waiting
+    # for additional logs now?
+    waiting_for_logs = False
```

Comment:
Don't we always want to wait for logs?

Reply:
used here:

>         # if the execution ended, wait one iteration for additional logs
        if upload_ended:
            if waiting_for_logs:
                break
            else:
                waiting_for_logs = True

Reply:
Same question. Why wouldn't we want to wait for all logs to plot?